Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +16,70 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalAmphoursUsed_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if(currentNow_ < 0)
+        return true;
+    else
+        return false;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    double totalHoursRemaining, msRemaining;
+    int h, m, s, ms;
+
+    if(isCharging())
+        totalHoursRemaining = totalAmpHoursUsed() / currentNow_ * -1;
+    else
+        totalHoursRemaining = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / currentNow_;
+
+    //Checking for hour to be in between 0 and 24
+    while(totalHoursRemaining > 24)
+        totalHoursRemaining -= 24;
+
+    //Converting totalHoursRemaining into h, m, s, ms
+    h = floor(totalHoursRemaining);
+    msRemaining = (totalHoursRemaining - h) * 3600000;
+    m = floor(msRemaining / 60000);
+    msRemaining = msRemaining - (m * 60000);
+    s = floor(msRemaining / 1000);
+    msRemaining = msRemaining - (s * 1000);
```

Comment:
Consider making numbers that you use multiple times (i.e. 60 000 and 1000) into constant variables/macros.